File: django_prosemirror/schema/marks/link.py
"""Link mark definition."""

from prosemirror.model.schema import MarkSpec

from ...sanitize import is_safe_url, sanitize_url
from ..base import MarkDefinition


class LinkMark(MarkDefinition):
    """A link. Has href and optionally title attributes."""

    @property
    def name(self) -> str:
        return "link"

    def to_dom(self, mark, inline: bool) -> list:
        """Convert link mark to DOM representation."""
        # Documents stored before href validation existed may still hold an
        # executable URL, so neutralise rather than raise while rendering.
        base_attrs = {
            "href": sanitize_url(mark.attrs["href"]),
            "title": mark.attrs["title"],
        }
        attrs = self.class_mapping.apply_to_attrs(base_attrs, "link")
        return ["a", attrs, 0]

    def dom_matcher(self) -> list:
        """Return DOM parsing rules for link mark."""
        return [
            {
                "tag": "a",
                # Returning False drops the mark, keeping the text unlinked.
                "getAttrs": lambda attrs: (
                    {
                        "href": attrs.get("href"),
                        "title": attrs.get("title"),
                    }
                    if is_safe_url(attrs.get("href"))
                    else False
                ),
            }
        ]

    @property
    def spec(self) -> MarkSpec:
        return {
            "attrs": {"href": {}, "title": {"default": None}},
            "inclusive": False,
            "parseDOM": self.dom_matcher(),
            "toDOM": self.to_dom,
        }

File: frontend/schema/marks/link.ts
import {
    AttributeSpec,
    DOMOutputSpec,
    Mark,
    ParseRule,
} from "prosemirror-model";
import { type ClassMapping, MarkDefinition } from "@/schema/abstract";
import { MarkType } from "@/schema/types";
import { isSafeUrl, sanitizeUrl } from "@/utils/sanitize";

/**
 * Class that returns the spec of a code mark.
 * Prosemirror processes this mark like this ->
 * `<a href={attrs.href} title={attrs.title}>{ content }</a>`
 */
export class LinkMark extends MarkDefinition {
    /** Link mark. */
    constructor(classes: ClassMapping) {
        super(classes);
    }
    override name = MarkType.CODE;
    override inclusive?: boolean | undefined = false;
    override attrs?: { [name: string]: AttributeSpec } | undefined = {
        href: { validate: "string" },
        title: { default: null, validate: "string|null" },
    };
    override parseDOM: ParseRule[] = [
        {
            tag: "a[href]",
            getAttrs(dom) {
                const href = dom.getAttribute("href");
                // Returning false drops the mark, keeping the text unlinked.
                if (!isSafeUrl(href)) return false;
                return {
                    href,
                    title: dom.getAttribute("title"),
                };
            },
        },
    ];
    override toDOM(mark: Mark): DOMOutputSpec {
        const attrs = this.classMapping.apply_to_attrs(
            {
                // Documents stored before href validation existed may still
                // hold an executable URL, so neutralise rather than throw
                // while rendering.
                href: sanitizeUrl(mark?.attrs.href),
                title: mark?.attrs.title,
            },
            this.name,
        );
        return ["a", attrs, 0];
    }
}
